File: elixir/2559_vowel_strings.exs
# https://leetcode.com/problems/count-vowel-strings-in-ranges/
# 02/01/2025

defmodule Solution do
  @vowels MapSet.new(["a", "e", "i", "o", "u"])

  @spec vowel_strings(words :: [String.t()], queries :: [[integer]]) :: [integer]
  def vowel_strings(words, queries) do
    prefix_sum =
      words
      |> Enum.reduce({[0], 0}, fn word, {acc, count} ->
        new_count = count + if is_vowel_string?(word), do: 1, else: 0
        {[new_count | acc], new_count}
      end)
      |> Kernel.elem(0)
      |> Enum.reverse()
      |> List.to_tuple()

    queries
    |> Enum.map(fn [start, end_idx] ->
      Kernel.elem(prefix_sum, end_idx + 1) - Kernel.elem(prefix_sum, start)
    end)
  end

  defp is_vowel_string?(word) do
    [first, last] = [String.first(word), String.last(word)]
    MapSet.member?(@vowels, first) and MapSet.member?(@vowels, last)
  end
end

defmodule Main do
  def main do
    test_cases = [
      %{
        words: ["aba", "bcb", "ece", "aa", "e"],
        queries: [[0, 2], [1, 4], [1, 1]],
        expected: [2, 3, 0]
      },
      %{
        words: ["a", "e", "i"],
        queries: [[0, 2], [0, 1], [2, 2]],
        expected: [3, 2, 1]
      }
    ]

    Enum.each(test_cases, fn %{words: words, queries: queries, expected: expected} ->
      result = Solution.vowel_strings(words, queries)
      IO.puts("Input: words = #{inspect(words)}, queries = #{inspect(queries)}")
      IO.puts("Output: #{inspect(result)}")
      IO.puts("Expected: #{inspect(expected)}")
      IO.puts("Test #{if result == expected, do: "PASSED", else: "FAILED"}")
      IO.puts("------")
    end)
  end
end

Main.main()

File: rust/src/bin/2559_vowel_strings.rs
// https://leetcode.com/problems/count-vowel-strings-in-ranges/
// 02/01/2025

impl Solution {
    pub fn vowel_strings(words: Vec<String>, queries: Vec<Vec<i32>>) -> Vec<i32> {
        const VOWELS: [u8; 5] = [b'a', b'e', b'i', b'o', b'u'];
        let prefix_sum: Vec<i32> = words
            .iter()
            .scan(0, |state, w| {
                let bytes = w.as_bytes();
                let (first, last) = (bytes[0], bytes[bytes.len() - 1]);
                *state += (VOWELS.contains(&first) && VOWELS.contains(&last)) as i32;
                Some(*state)
            })
            .collect();

        queries
            .iter()
            .map(|q| {
                let (start, end) = (q[0] as usize, q[1] as usize);
                if start == 0 {
                    prefix_sum[end]
                } else {
                    prefix_sum[end] - prefix_sum[start - 1]
                }
            })
            .collect()
    }
}

#[derive(Debug)]
struct Solution {}

#[cfg(test)]
mod tests {
    use super::*;
    use rstest::rstest;

    #[rstest]
    #[case(
        ["aba", "bcb", "ece", "aa", "e"]
            .map(|s| s.to_string())
            .to_vec(),
        [[0, 2], [1, 4], [1, 1]]
            .map(|q| q.to_vec())
            .to_vec(),
        [2, 3, 0].to_vec(),
    )]
    #[case(
        ["a", "e", "i"]
            .map(|s| s.to_string())
            .to_vec(),
        [[0, 2], [0, 1], [2, 2]]
            .map(|q| q.to_vec())
            .to_vec(),
        [3, 2, 1].to_vec(),
    )]
    fn test_vowel_strings(
        #[case] words: Vec<String>,
        #[case] queries: Vec<Vec<i32>>,
        #[case] expected: Vec<i32>,
    ) {
        let result = Solution::vowel_strings(words, queries);
        assert_eq!(result, expected);
    }
}

#[cfg(not(test))]
fn main() {
    let test_cases = vec![
        (
            ["aba", "bcb", "ece", "aa", "e"]
                .map(|s| s.to_string())
                .to_vec(),
            [[0, 2], [1, 4], [1, 1]].map(|q| q.to_vec()).to_vec(),
            [2, 3, 0].to_vec(),
        ),
        (
            ["a", "e", "i"].map(|s| s.to_string()).to_vec(),
            [[0, 2], [0, 1], [2, 2]].map(|q| q.to_vec()).to_vec(),
            [3, 2, 1].to_vec(),
        ),
    ];

    for (words, queries, expected) in test_cases {
        let result = Solution::vowel_strings(words.clone(), queries.clone());
        println!(
            "Input: {:?}, Expected: {:?}, Got: {:?}",
            (words, queries),
            expected,
            result
        );
        assert_eq!(result, expected);
    }
}
